Add `Iterator::map_windows`

frank-king opened the pull request:
Tracking issue:  #87155.

This is inherited from the old PR  #82413.

Unlike #82413, this PR implements the `MapWindows` to be lazy: only when pulling from the outer iterator, `.next()` of the inner iterator will be called.

## Implementaion Steps
- [x] Implement `MapWindows` to keep the iterators' [*Laziness*](https://doc.rust-lang.org/std/iter/index.html#laziness) contract. 
- [x] Fix the known bug of memory access error.
- [ ] Full specialization of iterator-related traits for `MapWindows`.
    - [x] `Iterator::size_hint`,
    - [x] ~`Iterator::count`~,
    - [x] `ExactSizeIterator` (when `I: ExactSizeIterator`),
    - [x] ~`TrustedLen` (when `I: TrustedLen`)~,
    - [x] `FusedIterator`,
    - [x] ~`Iterator::advance_by`~,
    - [x] ~`Iterator::nth`~,
    - [ ] ...
- [ ] More tests and docs.

## Unresolved Questions:
- [ ] Is there any more iterator-related traits should be specialized? 
- [ ] Is the double-space buffer worth?
- [ ] Should there be `rmap_windows` or something else?
- [ ] Taking GAT for consideration, should the mapper function be `FnMut(&[I::Item; N]) -> R` or something like `FnMut(ArrayView<'_, I::Item, N>) -> R`? Where `ArrayView` is mentioned in https://github.com/rust-lang/generic-associated-types-initiative/issues/2. 
    - It can save memory, only the same size as the array window is needed,
    - It is more efficient, which requires less data copies,
    - It is possibly compatible with the GATified version of `LendingIterator::windows`.
    - But it prevents the array pattern matching like `iter.map_windows(|_arr: [_; N]| ())`, unless we extend the array pattern to allow matching the `ArrayView`.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @m-ou-se (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5437840515?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b9a19912ba9ef2f533094d3e1810e6e96d45cff8/checks/5437840515/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5438187061?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a656b7b411781ab3846296c822616ec06d6d8d79/checks/5438187061/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
i......................i......................i......................i.............................. 3700/3763
...............................................................
failures:

---- src/iter/traits/iterator.rs - iter::traits::iterator::Iterator::map_windows (line 1566) stdout ----
error: expected `,`, found `*`
   |
   |
40 |     .map_windows(|arr: &[_; 2] *arr);
   |                                ^ expected `,`

error: expected one of `#`, `&`, `(`, `-`, `...`, `..=`, `..`, `[`, `_`, `box`, `mut`, `ref`, `|`, identifier, or path, found `*`
   |
   |
40 |     .map_windows(|arr: &[_; 2] *arr);
   |                                ^ expected one of 15 possible tokens
error[E0433]: failed to resolve: use of undeclared type `NonFusedIterator`
  --> src/iter/traits/iterator.rs:1588:16
   |
   |
25 | let mut iter = NonFusedIterator::default();
   |                ^^^^^^^^^^^^^^^^ use of undeclared type `NonFusedIterator`
error[E0412]: cannot find type `NonFusedIterator` in this scope
  --> src/iter/traits/iterator.rs:1574:19
   |
   |
7  | struct NoneFusedIterator {
   | ------------------------ similarly named struct `NoneFusedIterator` defined here
11 | impl Iterator for NonFusedIterator {
   |                   ^^^^^^^^^^^^^^^^ help: a struct with a similar name exists: `NoneFusedIterator`

error[E0425]: cannot find value `state` in this scope
error[E0425]: cannot find value `state` in this scope
  --> src/iter/traits/iterator.rs:1579:12
   |
16 |         if state < 5 || state % 2 == 0 {

error[E0425]: cannot find value `state` in this scope
  --> src/iter/traits/iterator.rs:1579:25
   |
   |
16 |         if state < 5 || state % 2 == 0 {

error[E0425]: cannot find value `state` in this scope
  --> src/iter/traits/iterator.rs:1580:18
   |
```

</details>

frank-king commented on library/core/tests/iter/adapters/map_windows.rs:
```diff
@@ -0,0 +1,313 @@
+    {
+        // FIXME: SIGSEGV: invalid memory reference without `.filter_map(Some)`
+        let mut iter = iter.filter_map(Some).map_windows(|arr: &[_; N]| *arr);
+        // let mut iter = iter.map_windows(|arr: &[_; N]| *arr);
```
Now I'm confused here why if I remove `.filter_map(Some)`, it crashes with `SIGSEGV: invalid memory reference`.

frank-king replied:
It was caused by the wrong `.size_hint()`. When the buffer is initialized, the `.size_hint()` should be equal to the inner iterator's.

frank-king commented:
r? @rust-lang/libs
r? @LukasKalbertodt
r? @dtolnay

bors commented:
:umbrella: The latest upstream changes (presumably #95798) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #91970) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Mark-Simulacrum commented:
Going to mark as waiting-on-author since it's in draft and has some conflicts -- not really clear what the state of this PR is though. If you'd like a review, please leave a comment with `@rustbot ready`.

frank-king commented:
Got busy recently.

Actually, I think there should be more tests, while I'm not sure what are the best ways to test it. Maybe a review could help.

@rustbot ready

dtolnay commented:
Reassigning yaahc's reviews as she has stepped down from the review rotation.

r? rust-lang/libs-api

dtolnay commented:
Reassigning BurntSushi's reviews because based on `git log -i --grep=r=burntsushi` the last time they did rust-lang/rust reviews was over 3 years ago.

r? rust-lang/libs-api

dtolnay reviewed (commented):
The library API team already previously discussed and expressed approval of this exact API in https://github.com/rust-lang/rust/pull/82413#pullrequestreview-706596271, so this is good to land as an unstable API. For resolving doubts about tests (https://github.com/rust-lang/rust/pull/94667#issuecomment-1193419310), and to review and iterate on the implementation:

r? rust-lang/libs

In response to the questions listed at the bottom of https://github.com/rust-lang/rust/pull/94667#issue-1160563140: 

> Is there any more iterator-related traits should be specialized?

I wouldn't stress about this at this point. Trait impls can be added separately later if anyone notices one is missing. Please focus on the fundamentals (`next`...) being correct.

The tracking issue (#87155) would be the right place to make a note of what trait impls are outstanding.

> Is the double-space buffer worth?

Yes, this was our preference in https://github.com/rust-lang/rust/pull/82413#pullrequestreview-706596271. I'm sure we'll revisit again during stabilization.

> Should there be `rmap_windows` or something else?

I am skeptical that someone would be able to justify adding this.

> Taking GAT for consideration, should the mapper function be `FnMut(&[I::Item; N]) -> R` or something like `FnMut(ArrayView<'_, I::Item, N>) -> R`?

I'd definitely lean toward the array, but I expect we can hash this out during the period this is available on nightly. Copied to #87155.

frank-king commented:
Thank you for your review @dtolnay!

I have added some tests for other specialized methods.

Mark-Simulacrum commented on library/core/src/iter/adapters/map_windows.rs:
```diff
@@ -0,0 +1,379 @@
+    // `MapWindows::iter` (calling it the item iterator for convenient) into
+    // the buffer, but before that, it is uninitialized and thus `buffer` is
+    // `None`.
+    initialized: bool,
```
I think we should find a different name for this field, because whether the buffer is initialized or not can be tracked directly by the `Option`. Seeing a separate field for that makes me confused and it's not clear what exactly we're tracking here.

In practice I think the only reason we actually need a separate field is to track when the underlying iterator is *done*, so that we can avoid calling next on it repeatedly if there's Nones already interspersed. It seems like we could call this field `is_done` or store the underlying iterator in an Option as well? (I think wrapping the underlying iterator in Fuse could also work, but is trickier to implement since there's no is_done method on fused).

Regardless, we should update the comment on this field to reflect whatever we come up with.

frank-king replied:
To clarify, I list the three states of the iterator:
1. Uninitialized (`initialized == false && buffer.is_none()` in current code), before any `.next()` or `.advance_by()` called on the `MapWindows` iterator;
2. Initialized and active (`initialized == true && buffer.is_some()` in current code), since `.next()` or `.advance_by` called but not fused yet;
3. Fused (`initialized == true && buffer.is_none()` in current code), since the inner iterator yields a `None` and fused by us.

Your proposal is to distinguish states 1 & 3 with `is_done`, making it work similarly like `Fuse`. I think we may not need `is_done` if we wrap the `iter: I` with `iter: Fuse<I>` because everywhere we need `is_done` is to call `.next()`, `.advance_by()`, or `.size_hint()` etc. which is already wrapped in `Fuse`.

frank-king replied:
I'd like to add an `is_done` to `Fuse` and that will help to track this state here.

frank-king replied:
Oh, I realize what you have said. You meant that `.next()` would invoke the mapper function `f`, but `advance_by` would skip those mapping. They are unequivalent, is it?

Mark-Simulacrum commented on library/core/src/iter/adapters/map_windows.rs:
```diff
@@ -0,0 +1,379 @@
+                    Buffer::try_from_iter(&mut self.iter)
+                        .map_err(add_offset(steps))
+                        .map_err(exclude_first_window)?,
+                );
```
Something feels a little off here - I think we would expect that we successfully initialize and (eventually) return a Buffer if it's elements would be [x, x, x, to_be_returned] where x is an element currently advance_by'd before we start accumulating into the buffer. But I think this code would drop those x elements in the self.iter.advance_by, and so we would not match behavior with calling next() n times.

But I'm not quite sure, this feels tricky - I think we want a test at minimum that this behaves as expected, maybe you can point me to an existing one or add one.

Taking a step back, I think the thing that's making me feel weird is that we're only subtracting 1 from `n` to compute steps, not `N` or `N-1`, which would allow us to fill the buffer with the tail, as seems necessary for yielding the `n + 1`th element.

frank-king replied:
> But I think this code would drop those x elements in the self.iter.advance_by, and so we would not match behavior with calling next() n times.

I'm a little confused. Do you mean that whether those `x` elements (in `[x, x, x, to_be_returned]`) are yielded by `self.iter.next()` for `n - 1` times or by `self.iter.advance_by(n - 1)` differs because the former yields and drops those elements but the latter just skips them in place?

frank-king replied:
> But I'm not quite sure, this feels tricky - I think we want a test at minimum that this behaves as expected, maybe you can point me to an existing one or add one.

There is a test of `advance_by`: [link](https://github.com/rust-lang/rust/pull/94667/files#diff-82642196d9d615b4dcb26384c7d7d8f4a66746358409a21ad27c9926a034f9bdR312).

It tests the `advance_by` called at the very beginning before any `.next()` calls, and also after some `.next()` calls.

frank-king replied:
> Taking a step back, I think the thing that's making me feel weird is that we're only subtracting 1 from `n` to compute steps, not `N` or `N-1`, which would allow us to fill the buffer with the tail, as seems necessary for yielding the `n + 1`th element.

Subtracting 1 from `n` is because what we would made for the next window is `n..n + N`th elements (counted from zero), and `Buffer::try_from_iter` should count for one `.next()`, so we need to subtract 1.

frank-king replied:
Oh, I think using `Buffer::try_from_iter(self.iter.by_ref().skip(n - 1))` makes more sense here.

frank-king replied:
Oh, I realize what you have said. You meant that `.next()` should invoke the mapping function `f`, but we would skip that by calling `.advance_by` of the inner iterator. They are not equivalent, right?

I think we'd better remove the overwritten `.advance_by` method and leave it the default implementation.

Mark-Simulacrum commented on library/core/src/iter/adapters/map_windows.rs:
```diff
@@ -0,0 +1,379 @@
+
+        // SAFETY: the index is valid and this is element `a` in the
+        // diagram above and has not been dropped yet.
+        unsafe { (*to_drop).assume_init_drop() };
```
This is probably better as a ptr::drop_in_place, so that we don't need to move the element.

frank-king replied:
Thanks, will do.

Mark-Simulacrum commented on library/core/src/iter/adapters/map_windows.rs:
```diff
@@ -0,0 +1,379 @@
+}
+
+#[unstable(feature = "iter_map_windows", reason = "recently added", issue = "87155")]
+unsafe impl<I, F, R, const N: usize> TrustedLen for MapWindows<I, F, N>
```
The comment on [TrustedLen](https://doc.rust-lang.org/nightly/core/iter/trait.TrustedLen.html) to me implies that we can't implement it here. If the underlying iterator is infinitely long, then map_windows will return -(N-1) elements, right? But that is not an accurate count, just like Skip cannot be accurate in that case.

frank-king replied:
Yes, indeed, I didn't notice that comment.

Mark-Simulacrum commented:
@rustbot author

frank-king commented:
I removed the overwritten implementations of `advance_by`, `nth` and `count`, in order not to skip the side effects of invoking the mapping function.

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #106934) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Mark-Simulacrum commented on library/core/src/iter/adapters/fuse.rs:
```diff
@@ -24,6 +24,10 @@ impl<I> Fuse<I> {
     }
+    /// Whether the iterator is fused (i.e. it will return `None` forever)
+    pub(in crate::iter) fn is_fused(&self) -> bool {
+        self.iter.is_none()
```
This isn't going to work due to the FusedIterator optimization (see comment on `iter` above), I'm not sure there's an easy way to detect what you want via Fuse itself.

frank-king replied:
I agree. We don't know whether the inner iterator is fused if itself is a `FusedIterator`.

Mark-Simulacrum commented:
I think Fuse::is_fused won't quite work unfortunately, but other than that this looks good to me. Apologies for the slow review.

@rustbot author

frank-king committed:
Add Iterator::map_windows

This is inherited from the old PR.

This method returns an iterator over mapped windows of the starting
iterator. Adding the more straight-forward `Iterator::windows` is not
easily possible right now as the items are stored in the iterator type,
meaning the `next` call would return references to `self`. This is not
allowed by the current `Iterator` trait design. This might change once
GATs have landed.

The idea has been brought up by @m-ou-se here:
https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Iterator.3A.3A.7Bpairwise.2C.20windows.7D/near/224587771

Co-authored-by: Lukas Kalbertodt <lukas.kalbertodt@gmail.com>

frank-king commented:
Thanks for your review, and apologies for the slow reply, too. I've removed the `Fuse<I>` and replaced it with an `Option<I>` instead.

@rustbot ready

Emilgardis commented on library/core/src/iter/adapters/map_windows.rs:
```diff
@@ -0,0 +1,293 @@
+struct MapWindowsInner<I: Iterator, const N: usize> {
+    // We fuse the inner iterator because there shouldn't be "holes" in
+    // the sliding window. Once the iterator returns a `None`, we make
+    // our `MapWindows` iterator returns `None` forever.
```
typo
```suggestion
    // our `MapWindows` iterator return `None` forever.
```

Emilgardis commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -1591,6 +1592,160 @@ pub trait Iterator {
         Flatten::new(self)
     }
 
+    /// Calls the given function `f` for each contiguous window of size `N` over
+    /// `self` and returns an iterator over the outputs of `f`.
+    ///
+    /// In the following example, the closure is called three times with the
+    /// arguments `&['a', 'b']`, `&['b', 'c']` and `&['c', 'd']` respectively.
```
maybe this should mention overlap, like `slice::windows()` https://doc.rust-lang.org/std/primitive.slice.html#method.windows

Emilgardis commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -1591,6 +1592,160 @@ pub trait Iterator {
+    /// Note that the const parameter `N` is usually inferred by the
+    /// destructured argument in the closure.
+    ///
+    /// The returned iterator yields 𝑘 − `N` + 1 items (where 𝑘 is the number of
+    /// items yielded by `self`). If `self` yields fewer than `N` items, the
+    /// iterator returned from this method is empty.
```
```suggestion
    /// The returned iterator yields 𝑘 − `N` + 1 items (where 𝑘 is the number of
    /// items yielded by `self`). If `k` is less than `N`, this
    /// method yields an empty iterator.
```

Emilgardis commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -1591,6 +1592,160 @@ pub trait Iterator {
+    /// items yielded by `self`). If `self` yields fewer than `N` items, the
+    /// iterator returned from this method is empty.
+    ///
+    /// The returned iterator is always fused, because once `self` returns `None`,
+    ///  even if it returns a `Some(T)` again in the next iteration, we cannot put
+    /// it into a contigious array buffer, and thus the returned iterator is also
+    /// fused.
```
this is just an explanation of what a fuse is, maybe better to link to `FusedIterator`?

Emilgardis reviewed (commented):
some doc suggestions

Mark-Simulacrum commented:
@bors r+ rollup

Going to go ahead and approve this, but if you can follow up with any doc comments etc that makes sense to me. (This has been pending for long enough that I don't want to block on those minor changes).

bors commented:
:pushpin: Commit ca17752c448c3fce87eb3828a8388f2a21933416 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum ca17752c448c3fce87eb3828a8388f2a21933416 -->
<!-- homu: {"type":"Approved","sha":"ca17752c448c3fce87eb3828a8388f2a21933416","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

frank-king commented:
Thank you for your reviews @Emilgardis and @Mark-Simulacrum! I've revised the docs and rebased all commits into a single one.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit f0f9d88de6b74592a9fb2bb67c9a3c64edfcb54c has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum f0f9d88de6b74592a9fb2bb67c9a3c64edfcb54c -->
<!-- homu: {"type":"Approved","sha":"f0f9d88de6b74592a9fb2bb67c9a3c64edfcb54c","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GuillaumeGomez commented:
It failed in https://github.com/rust-lang/rust/pull/111760. wasm doesn't support unwinding panics.

@bors r-

Emilgardis commented:
The fix for that is adding `#[cfg(not(panic = "abort"))]` to the tests that need unwinding

frank-king commented:
@rustbot ready

Mark-Simulacrum commented:
r=me with a rebase

@rustbot author

frank-king commented:
Didn't see that message, sorry for the long wait.

r? @Mark-Simulacrum

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 97c953f561e564a1b4fee2570dd2df704423d287 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 97c953f561e564a1b4fee2570dd2df704423d287 -->
<!-- homu: {"type":"Approved","sha":"97c953f561e564a1b4fee2570dd2df704423d287","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 97c953f561e564a1b4fee2570dd2df704423d287 with merge 1b198b3a196442e14fb06978166ab46a4618d131...
<!-- homu: {"type":"BuildStarted","head_sha":"97c953f561e564a1b4fee2570dd2df704423d287","merge_sha":"1b198b3a196442e14fb06978166ab46a4618d131"} -->
